tests/test_promotion.py

```
@@ -0,0 +1,88 @@
+        _mock_registry(monkeypatch, 404)
+
+        with pytest.raises(ValueError, match="not found"):
+            promotion.promote("staging", "web", "9.9.9")
+        generator.return_value.generate.assert_not_called()
+
+    def test_rejects_non_platform_app(self, env_patched) -> None:
+        with pytest.raises(ValueError, match="not a platform app"):
+            promotion.promote("staging", "errors", "1.1.1")
+
+    def test_rejects_dev_env(self, env_patched) -> None:
+        with pytest.raises(ValueError, match="staging|prod"):
```

Ruff (0.15.15)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)